pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -27,65 +27,133 @@ import (
+	n.recalculateDaemonResources()
 
 	for _, it := range instanceTypeOptions {
+		// If a zero-resource pod can't fit, don't consider this instance type.  This occurs if the node overhead +
+		// daemon set requests is already larger than the instance type can support
+		if !n.newPodCanFit(nil, it) {
+			continue
+		}
+
 		// pre-filter our list of all possible instance types by what the provisioner allows
 		if !cloudprovider.Compatible(it, constraints.Requirements) {
 			continue
 		}
 		n.InstanceTypeOptions = append(n.InstanceTypeOptions, it)
 	}
-	// technically we might create some invalid construct here if the pod can't be supported by any instance type
-	// but this is ok as we just won't have any valid instance types, this pod won't be schedulable and nothing else
-	// will be compatible with this node
+
 	for _, p := range pods {
 		n.Add(p)
 	}
-	return n
+	if len(n.InstanceTypeOptions) == 0 {
+		return nil, errors.New("no instance type satisfied requirements")
+	}
+	return n, nil
 }
 
 func (n Node) Compatible(pod *v1.Pod) error {
-	if n.unschedulable {
-		return errors.New("node is unschedulable")
-	}
-
 	podRequirements := v1alpha5.NewPodRequirements(pod)
-	if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
+	if err := n.Requirements.Compatible(podRequirements); err != nil {
 		return err
 	}
 
-	tightened := n.Constraints.Requirements.Add(podRequirements.Requirements...)
+	tightened := n.Requirements.Add(podRequirements.Requirements...)
 	// Ensure that at least one instance type of the instance types that we are already narrowed down to based on the
 	// existing pods can support the pod resources and combined pod + provider requirements
+	podRequests := resources.RequestsForPods(pod)
 	for _, it := range n.InstanceTypeOptions {
-		if cloudprovider.Compatible(it, tightened) && n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
+		newSize := resources.Merge(n.reservedResources(it), podRequests)
+		if cloudprovider.Compatible(it, tightened) &&
+			n.newPodCanFit(newSize, it) &&
+			n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
 			return nil
 		}
 	}
 	return errors.New("no matching instance type found")
 }
 
+func (n Node) reservedResources(it cloudprovider.InstanceType) v1.ResourceList {
+	return resources.Merge(it.Overhead(), n.daemonResources, n.podResources)
+}
+
+func (n *Node) newPodCanFit(newSize v1.ResourceList, it cloudprovider.InstanceType) bool {
+	for resourceName, totalQuantity := range it.Resources() {
+		reservedQuantity := newSize[resourceName]
+		if reservedQuantity.Cmp(totalQuantity) > 0 {
+			return false
+		}
+	}
+
+	instancePodMax := it.Resources()[v1.ResourcePods]
+	if !instancePodMax.IsZero() && instancePodMax.CmpInt64(int64(len(n.Pods)+1)) < 0 {
+		return false
+	}
+	return true
+}
+
+func (n *Node) RequiredResources() v1.ResourceList {
```

Comment:
If this is only used by logging, WDYT about making node a Stringer and just log `launching node %s`

Reply:
Modified node to have a String() method and limited instance types to 5:

```
karpenter-69b966c4df-2jdh8 controller 2022-03-27T02:18:36.443Z	INFO	controller.provisioning	Launched instance: i-0db3804e44367a4ed, hostname: ip-192-168-135-166.us-west-2.compute.internal, type: t3a.2xlarge, zone: us-west-2c, capacityType: on-demand	{"commit": "c456cf5", "provisioner": "default"}
karpenter-69b966c4df-2jdh8 controller 2022-03-27T02:18:36.443Z	INFO	controller.provisioning	Launched with 5 pods using resources cpu: 5315m memory: 1093Mi ephemeral-storage: 5Gi  from types c1.xlarge, c3.2xlarge, c4.2xlarge, c5ad.2xlarge, c6i.2xlarge and 205 others	{"commit": "c456cf5", "provisioner": "default"}
karpenter-69b966c4df-2jdh8 controller 2022-03-27T02:18:36.536Z	INFO	controller.provisioning	Bound 5 pod(s) to node ip-192-168-135-166.us-west-2.compute.internal	{"commit": "c456cf5", "provisioner": "default"}
```